docs/src/architecture/domain-model.md

```diff
@@ -0,0 +1,214 @@
+  * On a **PLC** it's a _control application_, created by an _automation engineer_ in a domain specific graphical programming language (like "Ladder Diagram" or "Function block diagram")
+  * Or on a **SoC-based** or **Microcontroller-based** system it's an _application program_, created by an _embedded software engineer_ usually in C/C++
+* The **OS / Libs / Runtime** provide basic functionality to the **Domain Application**
+  <!-- TODO: add somehow "used to abtracts the hardware. But: on a microcontroler usually less abtraction, more hw dep on the domain app, and even no OS" -->
+* **Sensors** and **Actuators** enable the **device** to observe and control physical behaviour on the shopfloor or device's environment.
+  * can be integrated in the **device's** hardware,
+    or connected to the **device** via some **Fieldbus** (e.g. Modbus, CANopen, PROFINET, EtherCAT, ...) or
+    some **Local Interface** as USB, UART, SPI, I2C, ...
+  * can be simple peripherals as a light barrier, or a sensor for temperature or pressure;
+    or complex equipments as robotic arms or even other **devices**
+* **Inputs / Outputs** are the communication channels between the **Domain Application** and **Sensors** and **Actuators**
+  * drivers (as part of the **OS / Libs / Runtime** and/or the **Domain Application**) do expose all data from
+    **Sensors** and **Actuators** to the **device** as inputs or outputs
+  * also the **Domain Application** can expose data as input or output (e.g. own _signals_ or _states_)
+
+## IoT Cloud Domain
+
+Cloud end-points could be IoT Cloud Platforms, as e.g. [Cumulocity](https://www.softwareag.cloud/site/product/cumulocity-iot.html), [Azure](https://azure.microsoft.com) or [AWS](https://aws.amazon.com); or end-points for specific IoT tasks, as e.g. a time-series database to collect measurements.
+
+* Each **device** is represented in the IoT cloud by an individual **device twin**
+* All information / data or functionality the device provides to the cloud appears in context of it's **device twin**
+* The focus of **thin-edge** is on two IoT Cloud aspects:
+  - 1) **Telemetry Data Handling**
+  - 2) **Device Management**
+
+### Telemetry Data Handling
+
+**Telemetry data** are the measurements, events and alarms collected or raised by sensors, actuators or domain applications.
+
+  * **Measurement**, is a single value or set of values
+    * could be be a mix of values produced by one or more sensors and values calculated by the device's domain application
+    * values could be a mix of numbers, strings or booleans
+    * has one timestamp
+    * releates to one **Metric**
+  * **Metric**, is a time-series of measurements
+    * contains a _source device_
+    * contains a _type name_
+    * optionally contains _units_ for the **measurements**
+  * TODO: Example for metric and it's measurements
+  * **Command**, is a single value or set of values
```

Comment:
1) Having "Command" under "telemetry" heading feels a bit strange to me. Isn't telemetry "device to cloud" only?
2) IMHO: "A command is a cloud orginated request to change the state of the device".

Reply:
w.r.t. Point 1, yes technically control would be [telecommand](https://en.wikipedia.org/wiki/Telecommand). We could rename the property to something more generic e.g. `communication_descriptor`, or `device_descriptor`

Reply:
w.r.t. Point 2, I wouldn't limit a command just coming from the cloud. There are some instances where supporting a locally triggered command would is useful. And since the operation interface is driven by the local mqtt broker, then there is nothing stopping other services from triggering the same commands.

Use-case 1: A user is connected via ssh to a device and would like to check the handling of operations and could trigger a cli command to trigger the same process in the agent.

Use-case 2: There is a complex setup of devices with child devices where the firmware version needs to match on both the parent and child. When a child device is replaced, there is some additional service running which makes the decision which firmware version should be applied to the newly replaced device. The service then issues the command to the mqtt broker.

Reply:
I see the term `command` is really missleading here.
The original aim for the `command` was to have a kind of birectional communication for measurements.
  * Most motivation for that was the concept of "setpoints" from OT area
    * e.g. upper limit and lower limit of a climate control
  * A 'setpoint' as I would describe it:
    * has a value that is visible in the cloud, similar to a measurement (e.g. a "temperature upper limit")
    * is located and used on the device-site by the control-logic
    * it's value can be manipulated from cloud-site
      * a manipulation from cloud leads not to an immediate action (like an operation), 
      but the new value is used by the control application from next control-cycle on
    * and, it's value can also be manipulated from the device-site 

From my POV that is a concept missing in C8Y. However with **thin-edge** we should consider it. For C8Y we might able to emulate it somehow with C8Y operations. For other clouds it can be mapped to whatever the cloud supports.

However as already said on top, the term `command` is missleading. Instead I prefer to name it `setpoint`.
@vaillant , @reubenmiller , what do you think about that?